File: app/src/main/java/com/mcal/disassembler/view/FloatingMenuView.java
package com.mcal.disassembler.view;

import android.annotation.SuppressLint;
import android.content.ClipDescription;
import android.content.ClipboardManager;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.appcompat.widget.AppCompatImageButton;

import com.mcal.disassembler.R;
import com.mcal.disassembler.nativeapi.DisassemblerSymbol;
import com.mcal.disassembler.nativeapi.Searcher;

import java.util.Vector;

@SuppressLint("ViewConstructor")
public class FloatingMenuView extends RelativeLayout {
    private final Context context;
    private final EditText editText;
    private final TextView text;
    private final String path;

    FloatingMenuView(Context c, FloatingMenu menu, String filePath, int width, int height) {
        super(c);
        this.context = c;
        this.path = filePath;
        View view = LayoutInflater.from(c).inflate(R.layout.floating_menu, null);

        int buttonWidth = width / 6;

        AppCompatImageButton imageButtonClose = (AppCompatImageButton) view.findViewById(R.id.floatingmenuButtonClose);
        imageButtonClose.setOnClickListener(p1 -> {
            menu.dismiss();
            new FloatingButton(context, path).show();
        });
        setLayoutParams(imageButtonClose, buttonWidth);

        AppCompatImageButton imageButtonHide = (AppCompatImageButton) view.findViewById(R.id.floatingmenuButtonHide);
        imageButtonHide.setOnClickListener(p1 -> menu.dismiss());
        setLayoutParams(imageButtonHide, buttonWidth);
        editText = (EditText) view.findViewById(R.id.floatingmenuEditText);
        editText.postDelayed(() -> {
            InputMethodManager keyboard = (InputMethodManager) context.getSystemService(Context.INPUT_METHOD_SERVICE);
            keyboard.showSoftInput(editText, 0);
        }, 50);


        buttonWidth *= 0.75;
        AppCompatImageButton imageButtonClear = (AppCompatImageButton) view.findViewById(R.id.floatingmenuButtonClear);
        imageButtonClear.setOnClickListener(p1 -> editText.getText().clear());
        setLayoutParams(imageButtonClear, buttonWidth);

        AppCompatImageButton imageButtonPaste = (AppCompatImageButton) view.findViewById(R.id.floatingmenuButtonPaste);
        imageButtonPaste.setOnClickListener(p1 -> {
            editText.setText(readFromClipboard());
        });
        setLayoutParams(imageButtonPaste, buttonWidth);

        AppCompatImageButton imageButtonSearch = (AppCompatImageButton) view.findViewById(R.id.floatingmenuButtonSearch);
        imageButtonSearch.setOnClickListener(p1 -> search(editText.getText().toString()));
        text = (TextView) view.findViewById(R.id.floatingmenuTextView);
        setLayoutParams(imageButtonSearch, buttonWidth);

        AppCompatImageButton imageButtonCopy = (AppCompatImageButton) view.findViewById(R.id.floatingmenuButtonCopy);
        imageButtonCopy.setOnClickListener(p1 -> {
            String stringYouExtracted = text.getText().toString();
            int startIndex = text.getSelectionStart();
            int endIndex = text.getSelectionEnd();
            stringYouExtracted = stringYouExtracted.substring(startIndex, endIndex);

            android.content.ClipboardManager clipboard = (android.content.ClipboardManager) context.getSystemService(Context.CLIPBOARD_SERVICE);
            android.content.ClipData clip = android.content.ClipData.newPlainText("Copied Text", stringYouExtracted);
            clipboard.setPrimaryClip(clip);
        });
        setLayoutParams(imageButtonCopy, buttonWidth);

        this.addView(view);
    }

    public String readFromClipboard() {
        ClipboardManager clipboard = (ClipboardManager) context.getSystemService(Context.CLIPBOARD_SERVICE);
        if (clipboard.hasPrimaryClip()) {
            android.content.ClipDescription description = clipboard.getPrimaryClipDescription();
            android.content.ClipData data = clipboard.getPrimaryClip();
            if (data != null && description != null && description.hasMimeType(ClipDescription.MIMETYPE_TEXT_PLAIN))
                return String.valueOf(data.getItemAt(0).getText());
        }
        return null;
    }

    private void search(String name) {
        try {
            String localText = new String();
            Vector<DisassemblerSymbol> symbols = new Vector<DisassemblerSymbol>();
            symbols = Searcher.search(name);
            for (DisassemblerSymbol symbol : symbols)
                localText = localText + symbol.getDemangledName() + "\n";
            text.setText(localText);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void setLayoutParams(View view, int w) {
        ViewGroup.LayoutParams params = view.getLayoutParams();
        params.width = w;
        view.setLayoutParams(params);
    }
}

File: app/src/main/java/com/mcal/disassembler/view/FloatingMenuView.kt
package com.mcal.disassembler.view

import android.annotation.SuppressLint
import android.app.Activity
import android.graphics.Color
import android.text.Editable
import android.text.TextUtils
import android.text.TextWatcher
import android.view.WindowManager
import androidx.core.app.ActivityCompat
import com.mcal.disassembler.R
import com.mcal.disassembler.activities.BaseActivity
import com.mcal.disassembler.adapters.SymbolsItem
import com.mcal.disassembler.data.Preferences
import com.mcal.disassembler.databinding.FloatingMenuBinding
import com.mcal.disassembler.nativeapi.Dumper
import com.mcal.disassembler.utils.StringHelper
import com.mikepenz.fastadapter.FastAdapter
import com.mikepenz.fastadapter.adapters.ItemAdapter

@SuppressLint("ViewConstructor")
class FloatingMenuView internal constructor(
    private val activity: Activity,
    menu: FloatingMenu,
    private val path: String,
) : SymbolsSearchView(
    activity
) {
    private val binding by lazy { FloatingMenuBinding.inflate(activity.layoutInflater) }
    private val itemAdapter = ItemAdapter<SymbolsItem>()
    private val fastAdapter = FastAdapter.with(itemAdapter)
    private val params =
        WindowManager.LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT)

    init {
        val recyclerView = binding.recyclerView.apply {
            adapter = fastAdapter
        }

        if (data.isEmpty()) {
            initData()
        }

        val searchText = binding.search
        val clearBtn = binding.clearText.apply {
            setOnClickListener {
                searchText.setText("")
            }
        }
        searchText.addTextChangedListener(object : TextWatcher {
            override fun onTextChanged(
                s: CharSequence,
                start: Int,
                before: Int,
                count: Int
            ) = Unit

            override fun beforeTextChanged(
                s: CharSequence,
                start: Int,
                count: Int,
                after: Int
            ) = Unit

            override fun afterTextChanged(s: Editable) {
                clearBtn.visibility = if (s.isEmpty()) {
                    GONE
                } else {
                    VISIBLE
                }
                if (canStartFilterProcess) {
                    if (!TextUtils.equals(s, lastValue)) {
                        val constraint = s.toString()
                        lastValue = constraint
                        recyclerView.scrollToPosition(0)
                        canStartFilterProcess = false
                        filter(constraint)
                        return
                    }
                    return
                }
                newValue = s.toString()
            }
        })

        binding.floatingmenuButtonClose.apply {
            setOnClickListener {
                menu.dismiss()
                FloatingButton(activity, path).show()
            }
        }
        binding.floatingmenuButtonHide.apply {
            setOnClickListener {
                menu.dismiss()
            }
        }
        binding.floatingmenuButtonPaste.setOnClickListener {
            binding.search.setText(StringHelper.readFromClipboard(activity))
        }
        val preferences = Preferences(context)
        binding.regex.setBackgroundColor(
            if (preferences.regex) ActivityCompat.getColor(
                context,
                R.color.colorAccent
            ) else {
                Color.TRANSPARENT
            }
        )
        binding.regex.setOnClickListener {
            if (preferences.regex) {
                preferences.regex = false
                binding.regex.setBackgroundColor(Color.TRANSPARENT)
            } else {
                preferences.regex = true
                binding.regex.setBackgroundColor(
                    ActivityCompat.getColor(
                        context,
                        R.color.colorAccent
                    )
                )
            }
        }
        addView(binding.root, params)
    }

    private fun updateAdapter(list: MutableList<Map<String, Any>>) {
        var item: Map<String, Any>
        for (i in list.indices) {
            item = list[i]
            itemAdapter.add(
                SymbolsItem()
                    .withContext(activity)
                    .withId(i.toLong())
                    .withIcon(item["img"] as Int)
                    .withTitle(item["title"] as String)
                    .withSubTitle(item["info"] as String)
                    .withSymbolType(item["type"] as Int)
                    .withPath(path)
            )
        }
        BaseActivity.setVisibility(binding.progress, GONE)
    }

    private fun initData() {
        BaseActivity.setVisibility(binding.progress, VISIBLE)
        val list = symbolsFilteredList
        if (list.isNotEmpty()) {
            list.clear()
        }
        var map: MutableMap<String, Any>
        for (i in Dumper.symbols.indices) {
            map = HashMap()
            when (Dumper.symbols[i].type) {
                1 -> map["img"] = R.drawable.ic_box_blue
                2 -> map["img"] = R.drawable.ic_box_red
                else -> map["img"] = R.drawable.ic_box_green
            }
            map["title"] = Dumper.symbols[i].demangledName
            map["info"] = Dumper.symbols[i].name
            map["type"] = Dumper.symbols[i].type
            list.add(map)
        }
        list.sortBy {
            it["title"] as String
        }
        updateSymbolsSize(list)
        updateAdapter(list)
        val dataList = data
        if (dataList.isNotEmpty()) {
            dataList.clear()
        }
        dataList.addAll(list)
    }

    private fun updateSymbolsSize(list: ArrayList<Map<String, Any>>) {
        val symbolsSizeView = binding.symbolsSize
        val dataSize = list.size.toString()
        if (symbolsSizeView.text.toString() != dataSize) {
            symbolsSizeView.text = buildString {
                append(activity.getString(R.string.symbols_count))
                append(dataSize)
            }
        }
    }

    override fun onStartSearch() {
        BaseActivity.setVisibility(binding.progress, VISIBLE)
    }

    override fun onFoundApp(list: ArrayList<Map<String, Any>>, mode: Boolean) {
        BaseActivity.setVisibility(
            binding.symbolsNotFound, if (mode) {
                GONE
            } else {
                VISIBLE
            }
        )
        BaseActivity.setVisibility(
            binding.recyclerView, if (mode) {
                VISIBLE
            } else {
                GONE
            }
        )
        updateSymbolsSize(list)
        if (itemAdapter.adapterItemCount >= 0) {
            itemAdapter.clear()
        }
        updateAdapter(list)
    }
}